src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs

```diff
@@ -88,6 +90,120 @@ public void TestAsyncOutputStream()
             }
         }
 
+        [Fact]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.Uap, "Pipe doesn't work well on UAP")]
+        public void TestAsyncOutputStream_BeginCancelBegin_OutputReadLine()
+        {
+            using (AnonymousPipeServerStream pipeWrite = new AnonymousPipeServerStream(PipeDirection.Out, HandleInheritability.Inheritable))
+            using (AnonymousPipeServerStream pipeRead = new AnonymousPipeServerStream(PipeDirection.In, HandleInheritability.Inheritable))
+            {
+                var dataReceived = new List<int>();
+                var dataArrivedEvent = new AutoResetEvent(false);
+
+                using (Process p = CreateProcessPortable(TestAsyncOutputStream_BeginCancelBegin_OutputReadLine_RemotelyInvokable, $"{pipeWrite.GetClientHandleAsString()} {pipeRead.GetClientHandleAsString()}"))
+                {
+                    p.StartInfo.RedirectStandardOutput = true;
+                    p.OutputDataReceived += (s, e) =>
+                    {
+                        if (e.Data != null)
+                        {
+                            dataReceived.Add(int.Parse(e.Data));
+                            dataArrivedEvent.Set();
+                        }
+                    };
+
+                    // Start child process
+                    p.Start();
+
+                    pipeWrite.DisposeLocalCopyOfClientHandle();
+                    pipeRead.DisposeLocalCopyOfClientHandle();
+
+                    // Wait child process start
+                    Assert.True(WaitPipeSignal(pipeRead, WaitInMS), "Child process not started");
+
+                    //Start listening and produce output 1
+                    p.BeginOutputReadLine();
+                    pipeWrite.WriteByte(0);
+
+                    // Wait child signal produce number 1
+                    Assert.True(WaitPipeSignal(pipeRead, WaitInMS), "Missing child signal for value 1");
+                    Assert.True(dataArrivedEvent.WaitOne(WaitInMS), "Value 1 not received");
+
+                    //Stop listening and signal to produce value 2
+                    p.CancelOutputRead();
+                    pipeWrite.WriteByte(0);
+
+                    // Wait child signal produce number 2
+                    Assert.True(WaitPipeSignal(pipeRead, WaitInMS), "Missing child signal for value 2");
+                    // We need to sleep to be sure to drain async queue
```

Comment:
What happens if things are delayed and we don't actually completely drain the queue in this 500ms?  Will the test spuriously fail, or will the test simply not end up validating what it was intended to validate?  I'm ok with the latter, but we'll need to fix the former if that's the case.

Reply:
> Will the test spuriously fail, or will the test simply not end up validating what it was intended to validate?

Yes I'm aware of this and I expected your concern, but isn't too risky "not end up validating what it was intended to validate"? With future changes this test could became unuseful(fail forever), right?
I thought to wait for more time, classic 30 seconds or put test in a loop with some different wait time(to speed up test) , for instance wait 1,3,6,10 seconds and if fails every times fail test.
What do you think?

Reply:
> What do you think?

I expect we'd still see failures in CI, and if it fails, it won't be clear whether that's because there was a bug in the library or the test was just flaky, making the test irrelevant.

I'm not clear why the non-determinism is needed, though.  Seems like you could have two tests:
- One to validate that cancellation causes the resulting data to be thrown away
- One to validate that additional data arrives after reading is restarted

The former would synchronize between the parent and child by having the child send data and the parent read it all, then canceling and then signaling to the child to send more, then the child exiting, and the parent validating that it saw the stream end without having received additional data from the child.

The latter would do something similar.  The parent would read the data from the child, then call cancel, then signal for the child to send more, then start reading again, then signal to the child to send more, and then validate that it did in fact receive more... and we don't need to tell the difference between the data in the interim between the cancellation and the restart, and after the restart, because the first test validated we don't get data from the former.